### ISSUE TITLE
Exclude file is being ignored

### ISSUE THREAD

**Issue by cafuego** [restic/restic#1670]
After updating from restic 0.8.1 to 0.8.3 (with no other changes to my setup) restic has started throwing a permissions error on a path that is listed in an exclude file, so should be getting ignored.

## Output of `restic version`

```
restic 0.8.3
compiled with go1.10 on linux/amd64
```

## How did you run restic exactly?

```
restic-0.8.3 backup --exclude-file /path/to/exclude --no-cache -r gs:bucket:/restic /mnt/project
using parent snapshot e3f71c8b
scan [/mnt/project]
error for /mnt/project/.glusterfs: open /mnt/project/.glusterfs: permission denied
...
```

The exclude file contains:

```
.glusterfs
.glusterfs/*
.migrated
.migrated/*
.trashcan
.trashcan/*
css/*
ctools/*
js/*
styles/*
temp/*
tmp/*
xmlsitemap/*
```

## What backend/server/service did you use to store the repository?

Google Cloud

## Expected behavior

I want restic to not try to access the paths listed in the exclude file.

## Actual behavior

Restic tries to access files and directories under paths listed in the exclude file.

## Do you have any idea what may have caused this?

I updated from restic 0.8.1 to 0.8.3.

0.8.1 did not throw the permissions error on the .glusterfs directory when running backups and restoring a snapshot made with 0.8.1 does not restore a .glusterfs dir (or any of the other paths listed in the exclude file), so it was happy with my exclude file syntax.

## Do you have an idea how to solve the issue?

Other than downgrade to 0.8.1, sorry, no.

## Did restic help you or made you happy in any way?

Restic makes me happy, but not nearly as happy as the helpfulness and responsiveness of the developer :-)

**Comment by paulschreiber** [comment]
Should you be using `--exclude-file=/path/to/exclude` instead of `--exclude-file /path/to/exclude`?

**Comment by cafuego** [comment]
Perhaps, but the current syntax worked fine on 0.8.1. I would expect that to also not work if I got the syntax wrong.

**Comment by fd0** [comment]
Thanks for the report! I think this is a bug. Both will work, `--exclude-file=foo` and `--exclude-file foo`.

As far as I can remember, we haven't changed anything in the exclude handling code since 0.8.1.

If you know how to do a `git bisect` between 0.8.1 and master, that'd be a good starting point to find out where the change in behavior was introduced. Do you know how to do that?

For each tested commit, you then run restic with a local repo and only list the path in question for backup, like this:
```
$ backup --exclude-file /path/to/exclude --no-cache /mnt/project/.glusterfs
```
That should produce the same error.

**Comment by fd0** [comment]
For the record: I'm unable to reproduce it. Can you please check that neither the file/dir `/mnt/project/.glusterfs` nor the lines in the exclude file have any leading/trailing spaces?

**Comment by sonicdoe** [comment]
I’m running into a similar error, see https://github.com/restic/restic/issues/1706.

I’ve decided to open a separate issue because I was able to reproduce it in v0.8.1 and because it’s not limited to `--exclude-file`. Still, it might very well be the same underlying issue.

**Comment by fd0** [comment]
Ah, I only now saw your comment here. I think you're right, it's the same issue. I did not recognize the underlying cause before, thank you!

**Comment by fd0** [comment]
So, in summary: This issue is caused by restic trying to `stat` a directory before checking if it would be excluded. It's hard to change that with the way the code is structured right now. I'll try to get this one done after #1494 is finished. :)

**Comment by cafuego** [comment]
Ha! I only just got around to checking this yesterday and found I get the same access warning locally on 0.8.1 that I did on 0.8.3. Remotely I hadn't seen it on 0.8.1, so I was about to close this issue assuming I screwed up the invocation or listing of subdirs in the exclude file. Glad I didn't :-)

**Comment by sonicdoe** [comment]
This should be fixed with the new archiver code (https://github.com/restic/restic/pull/1494) in restic 0.9.0 🙂

**Comment by fd0** [comment]
Indeed, can somebody confirm?

**Comment by sonicdoe** [comment]
I did a quick test run and 1.9.0 no longer prints a “permission denied“ error for excluded folders.

restic 1.8.3:

```
$ mkdir /tmp/restic /tmp/restic/source /tmp/restic/destination
$ mkdir -m 000 /tmp/restic/source/verboten
$ restic init --repo /tmp/restic/destination
enter password for new repository: 
enter password again: 
created restic repository 085b3c76b9 at /tmp/restic/destination

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
$ restic --repo /tmp/restic/destination backup /tmp/restic/source --exclude=verboten
enter password for repository: 
password is correct
scan [/tmp/restic/source]
error for /tmp/restic/source/verboten: open /tmp/restic/source/verboten: permission denied
scanned 1 directories, 0 files in 0:00
[0:00]   0B / 0B  1 / 1 items  0 errors  ETA 0:00 
duration: 0:00
snapshot 40dc1520 saved
```

restic 1.9.0:

```
$ mkdir /tmp/restic /tmp/restic/source /tmp/restic/destination
$ mkdir -m 000 /tmp/restic/source/verboten
$ restic init --repo /tmp/restic/destination
enter password for new repository: 
enter password again: 
created restic repository 085b3c76b9 at /tmp/restic/destination

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
$ restic --repo /tmp/restic/destination backup /tmp/restic/source --exclude=verboten
enter password for repository: 
repository 085b3c76b9 opened successfully, password is correct

Files:           0 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added:      1.287 KiB

processed 0 files, 0 B in 0:02
snapshot 40dc1520 saved
```

**Comment by fd0** [comment]
Cool, thanks for the confirmation, I'm closing this issue.

**Comment by bjoe2k4** [comment]
close?

**Comment by cafuego** [comment]
I bumped my install to 0.8.3 last week and the warning has indeed gone away - thank you :-)

**Comment by fd0** [comment]
Yep, forgot to close. thanks everyone ;)
